### PR TITLE
Bridge: added free headers submission support to the substrate-relay

### PR DESCRIPTION
**Pull request by svyatonik**
Original PR: https://github.com/paritytech/parity-bridges-common/pull/2884. Since chain-specific code lives in the `parity-bridges-common` repo, some parts of original PR will require another PR

### CODE REVIEW

File: bridges/relays/client-substrate/src/test_chain.rs [hunk 1]
```diff
@@ -56,6 +56,7 @@ impl bp_runtime::Chain for TestChain {
 impl Chain for TestChain {
 	const NAME: &'static str = "Test";
 	const BEST_FINALIZED_HEADER_ID_METHOD: &'static str = "TestMethod";
+	const FREE_HEADERS_INTERVAL_METHOD: &'static str = "TestMethod";
```
Comment (acatangiu) [inline bridges/relays/client-substrate/src/test_chain.rs:59]: is it ok to have the same name for both these different methods?

Comment (svyatonik) [reply bridges/relays/client-substrate/src/test_chain.rs:59]: Yeah, it's just for tests and we do not have a mocked clients, which are able to respond to RPC calls => it is not tested

**Review by serban300 — APPROVED** [review]

File: bridges/relays/finality/src/finality_loop.rs [hunk 2]
```diff
@@ -302,17 +326,28 @@ impl<P: FinalitySyncPipeline, SC: SourceClient<P>, TC: TargetClient<P>> Finality
...
+		let selector = JustifiedHeaderSelector::new::<SC, TC>(
+			&self.source_client,
+			info,
+			self.sync_params.headers_to_relay,
+			free_headers_interval,
+		)
+		.await?;
 		// if we see that the header schedules GRANDPA change, we need to submit it
-		if self.sync_params.only_mandatory_headers {
+		if self.sync_params.headers_to_relay == HeadersToRelay::Mandatory {
```
Comment (bkontur) [inline bridges/relays/finality/src/finality_loop.rs:337]: What about allowing here also `HeadersToRelay::All`?

Comment (svyatonik) [reply bridges/relays/finality/src/finality_loop.rs:337]: Sorry, I don't get what you mean. Can you, please, elaborate? `All` and `Free` are handled later, in the `selector.select` call.

Comment (bkontur) [reply bridges/relays/finality/src/finality_loop.rs:337]: nvm, I need to walk through the entire PR first :)

**Pull request merged**